Switch to a 'main' branch name.

toonarmycaptain opened the pull request:
Allow Alan to comment on PR :)

mend-bolt-for-github[bot] committed:
Add .whitesource configuration file

toonarmycaptain committed:
Add miniconda_requisite

Try a miniconda install conditioned on checking conda --version
using the onfail key.

toonarmycaptain committed:
Update miniconda.sls with working install

Using requisite onfail, passing the conda --version command with the PATH that command should reference successfully if conda is installed correctly.

Because the PATH for miniconda is directly passed to the check_conda_install command, the step ensuring that miniconda is added to PATH needs to happen every time, as that state is not ensured by check_conda_install.

toonarmycaptain committed:
Add pip to python installs.

toonarmycaptain committed:
Add state installing Dask in python3.9 with pip.

toonarmycaptain committed:
Add draft conda_env_with_dask.sls

Creating the ev works. Activating it currently does not.

toonarmycaptain committed:
Merge pull request #1 from toonarmycaptain/whitesource/configure

Configure WhiteSource Bolt for GitHub

toonarmycaptain committed:
Source conda shell to activate conda env.

toonarmycaptain committed:
Add states to check/install dask in conda env.

toonarmycaptain committed:
Remove shell/env args, as sourcing shell directly

toonarmycaptain committed:
Comment unnecessary activation state.

Conda env must be activated within each state that uses the env.

toonarmycaptain committed:
Add comment indic check_conda_install origin.

alan-cugler commented on srv/salt/state/conda_env_with_dask.sls:
```diff
@@ -0,0 +1,33 @@
+    - shell: /bin/bash
+    - env:
+      - PATH: /root/miniconda3/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/root/bin
+    - require:
+      - check_conda_install  # miniconda.sls; will ensure conda is set up.
```
A comment or refactor is needed here to add context. doing this to much will make it harder for others to follow along. this one instance is fine with a comment.

alan-cugler replied:
actually is this necessary at all?

toonarmycaptain replied:
What context is needed? Is that comment sufficient? Would a full path to the state be better?

toonarmycaptain replied:
Wait, is it necessary?
My understanding is that, to some degree, states should be independent - so if a state relies on a prior condition being satisfied, or another state being true/complete, this should be explicit.

alan-cugler replied:
Being explicit isnt always necessary If you are coming from the assumption that something should already be on the system outside of salt then you can explicitly check, if its something you know your salt is installing in the highstate then you dont need to be explicit.

The other thing to consider is every time you reference another state or state file you raise the likelihood of a subsequent refactor forgetting to remove or adjust the referencing states that have been tethered to it. This eventually can make testing state a bit of a pain depending on your testing method, as simply removing a state temporarily might require lots of commenting out in other states.

toonarmycaptain replied:
Fair enough, I'll remove it. 

It *could* be added as an explicit `check_miniconda_is_installed_before_checking_for_env` but at that point, you could have many different states checking for the same things before doing anything, and I wouldn't want to go there.

toonarmycaptain replied:
Do I need a comment pointing out that this state presumes conda is installed at that specific path, or is it obvious?

alan-cugler commented on srv/salt/state/conda_env_with_dask.sls:
```diff
@@ -0,0 +1,33 @@
+    - name: 'conda env list | grep "py39_dask"'
+    - shell: /bin/bash
+    - env:
+      - PATH: /root/miniconda3/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/root/bin
```
does this replace the path thats on the system? what if I have other PATH variables present on the system?

toonarmycaptain replied:
I don't know. I presume so. 
The [salt cmd.run docs](https://docs.saltproject.io/en/latest/ref/states/all/salt.states.cmd.html) have:
```
{% set current_path = salt['environ.get']('PATH', '/bin:/usr/bin') %}

mycommand:
  cmd.run:
    - name: ls -l /
    - env:
      - PATH: {{ [current_path, '/my/special/bin']|join(':') }}
```
Which I believe would allow you to simply add something to PATH for the state to run. 
For what I am doing here, I don't know if overloading PATH, or adding to it would be universally helpful, in that I don't fully understand how order-precedence works with PATH. I can imagine a circumstance where simply augmenting PATH wouldn't use the path you're adding because another entry takes precedence.

toonarmycaptain replied:
Interesting, when I use 
```
check_conda_env_py39_dask:
  cmd.run:
    - name: 'conda create --name py39_dask python=3.9'
    - shell: /bin/bash
    - env:
      - PATH: {{ [current_path, '/root/miniconda3/bin']|join(':') }}
    - require:
      - check_conda_install  # miniconda.sls; will ensure conda is set up.
```
The command finds conda ok, but seems to not find the previously created env: `                  WARNING: A conda environment already exists at '/root/miniconda3/envs/py39_dask'
                  Remove existing environment (y/[n])?` Which suggests to me that sourcing conda.sh directly is the better way to go.

alan-cugler replied:
sourcing the conda file directly is... less elegant but I concede it might be better based on your findings.

toonarmycaptain replied:
I think the better systemic approach might ultimately be to have some sort of env var with the path to the folder where Miniconda is to be installed - pass that to the installer, and also to `'source {{ CONDA_DIR }}/etc/profile.d/conda.sh` that I'm using to bypass any `.bashrc` issues. Unless you make that whole path an env var: 
```
CONDA_DIR=~/here
CONDA_SHELL=CONDA_DIR/etc/profile.d/conda.sh
```
That would also have the flexibility to have multiple installations of conda on the system, not that I can imagine much of a need for that, but still.

alan-cugler commented on .whitesource:
```diff
@@ -0,0 +1,12 @@
+{
+  "scanSettings": {
+    "baseBranches": []
+  },
+  "checkRunSettings": {
+    "vulnerableCheckRunConclusionLevel": "failure",
+    "displayMode": "diff"
+  },
+  "issueSettings": {
+    "minSeverityLevel": "LOW"
+  }
+}
```
is this suppose to be here?

toonarmycaptain replied:
I've got whitesource enabled on the repo; it commits after the initial commit, but it's not relevant, no.

alan-cugler commented on srv/salt/state/conda_env_with_dask.sls:
```diff
@@ -0,0 +1,33 @@
+      - PATH: /root/miniconda3/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/root/bin
+    - onfail:
+      - cmd: check_conda_env_py39_dask
+
+activate_conda_env_py39_dask:
+  cmd.run:
+    - name: 'source ~/miniconda3/etc/profile.d/conda.sh && conda activate py39_dask'
+
+check_dask_install_in_ev_py39_dask:
+  cmd.run:
+    - name: 'source ~/miniconda3/etc/profile.d/conda.sh && conda activate py39_dask && conda list | grep "dask"'
+
```
This whole bit shouldnt be necessary unless some how conda force installs dask even if already present in its environment.

toonarmycaptain replied:
That's true, and I might have had that as an attempted workaround that didn't address a problem I was having.
I might also have been wary of [this comment](https://github.com/saltstack/salt/issues/52595#issuecomment-485509687) were  this state pattern to be used for many packages.
Just running `conda install dask` where it is already installed takes 2 full seconds (the check is 0.5s).

toonarmycaptain replied:
The activate env isn't necessary as it does nothing but activate the env. I'm not sure if I thought the activation persisted across states, or if it was an initial check that the env existed, but it serves no purpose.

alan-cugler commented on srv/salt/state/dask_in_python39_with_pip.sls:
```diff
@@ -0,0 +1,9 @@
+# File:/srv/salt/dask_in_python39_with_pip.sls
+
+install_dask_in_python39_with_pip:
+  pip.installed:
+    - names:
+      - dask
+    - bin_env: '/usr/bin/pip3.9'
+    - require:
+      - install_python39  # installs python39 and python39-pip
```
this is an independent salt state from the conda install dask I assume? otherwise its redundant and I do believe we can install dask with conda using pip.

toonarmycaptain replied:
Yes. This is independent.

alan-cugler commented on srv/salt/state/miniconda.sls:
```diff
@@ -1,26 +1,36 @@
+install_miniconda:
+  cmd.run:
+    - name: 'sudo /home/centos/Miniconda3-latest-Linux-x86_64.sh -b -u'
+    - onfail:
+      - cmd: check_conda_install
```
should be requiring ensure_minicond_installer instead in case the download fails.

alan-cugler commented on srv/salt/state/miniconda.sls:
```diff
@@ -1,26 +1,36 @@
 
-add_miniconda_to_bashrc:  # Add first, so that check in install_miniconda will work if conda is already installed.
+ensure_miniconda_path_in_bashrc:
   file.replace:
     - name: ~/.bashrc
     - pattern: /^export PATH="/root/miniconda3/bin:$PATH"/
     - repl: export PATH="/root/miniconda3/bin:$PATH"
     - append_if_not_found: True
```
still sexy, good job

toonarmycaptain replied:
Cheers. Apparently it was unnecessary, but like a lot of this, I've picked up some useful options in Salt, and flexed/developed some bash/*nix muscles :)

alan-cugler replied:
still useful example state, I actually went and found this code to demonstrate to a client yesterday

toonarmycaptain replied:
:)

> I find the miniconda.sls file needs renamed descriptors. you currently have 'miniconda_installed,' 'miniconda_installer_present,' 'install_miniconda.' These descriptors are almost like a tongue tie, I'd recommend refactoring to something like: "check_for_miniconda, download_miniconda, install_miniconda." Even though we are reusing miniconda over and over in the names, the rest of the descriptor is an action verb of that states purpose.

Is there some kinda of convention I'm misunderstanding? I understand the similar names issue, but I presume, with the paradigm of 'say what you want, not how to get there' that the states should reflect that tense. So, you "require miniconda is installed" rather than "require a check that miniconda is installed".

toonarmycaptain committed:
Source miniconda shell directly.

Eliminates need for messing with PATH or specifying bash shell.

This particular setup relies on there being conda.sh at the given path.

toonarmycaptain committed:
Refactor names to be more imperative.

toonarmycaptain committed:
Require install script before attempting to install.

toonarmycaptain committed:
Remove unnecessary .bashrc manipulation.

Sourcing ~/miniconda3/etc/profile.d/conda.sh renders this unnecessary.
Conda inserts what it needs in ~/.bashrc when it sets itself up.

toonarmycaptain committed:
Correct required state name.

toonarmycaptain committed:
Remove unneeded env activation state.

Conda env activation does carry between states.

alan-cugler commented on srv/salt/state/miniconda.sls:
```diff
@@ -1,26 +1,24 @@
-
-    - unless:
-        - cmd.run: echo $(conda --version)
+    - name: 'sudo /home/centos/Miniconda3-latest-Linux-x86_64.sh -b -u'
```
I think -u forces a install over the miniconda directory even if it exists. it might be better to not have that flag. Let me know if I got it wrong the purpose of that flag.

toonarmycaptain replied:
I can't actually find documentation for that `-u` flag. 
I did find [this](https://github.com/NanoComp/meep/issues/460#issuecomment-415569758) which suggests `-u` upgrades an existing installation. Whether it recognises that it is trying to install the already installed version and stops, idk.

Arguably, with the way the states are structured, if conda is being installed, it's because the checking state couldn't find conda, so either it's not there, or something is broken (whether conda itself, or the bash magic to access via `conda --version`), so forcing a reinstall via `-f` might be preferred in that case. Might have to specify the path with a `-p` flag if you want that, I haven't experimented. 

According to the [doc](https://docs.anaconda.com/anaconda/install/silent-mode/) the `-b` flag doesn't modify `~.bashrc` - which, although I think adding to the path automatically would be preferable, does allow the installer to processed without manual intervention (I tried and it hang without).

alan-cugler commented on srv/salt/state/miniconda.sls:
```diff
@@ -1,26 +1,24 @@
     - source: https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
     - skip_verify: True
-    - keep_source: True  # keep to avoid re-downloading.
+    - keep_source: True  # Keep to avoid re-downloading.
```
If a miniconda installation is suppose to be once and then always present, then this download is sitting taking up space. Also if enough time passes and its time to install and updated miniconda version, this might cause salt to use the older version miniconda installer rather than pulling down the newest one.

toonarmycaptain replied:
Fair enough, done.

alan-cugler reviewed (commented):
I find the miniconda.sls file needs renamed descriptors. you currently have 'miniconda_installed,' 'miniconda_installer_present,' 'install_miniconda.' These descriptors are almost like a tongue tie, I'd recommend refactoring to something like: "check_for_miniconda, download_miniconda, install_miniconda." Even though we are reusing miniconda over and over in the names, the rest of the descriptor is an action verb of that states purpose.

toonarmycaptain committed:
Do not cache miniconda install file.

toonarmycaptain committed:
Remove check for conda; it should already be installed.

Check relies on state in another file, which is fragile.

toonarmycaptain commented:
> I find the miniconda.sls file needs renamed descriptors. you currently have 'miniconda_installed,' 'miniconda_installer_present,' 'install_miniconda.' These descriptors are almost like a tongue tie, I'd recommend refactoring to something like: "check_for_miniconda, download_miniconda, install_miniconda." Even though we are reusing miniconda over and over in the names, the rest of the descriptor is an action verb of that states purpose.


Is there some kinda of convention I'm misunderstanding? I understand the similar names issue, but I presume, with the paradigm of 'say what you want, not how to get there' that the states should reflect that tense. So, you "require miniconda is installed" rather than "require a check that miniconda is installed".

toonarmycaptain committed:
Refactor state names to be less repetitive.
